windows/guest_customization/win_gosc_prepare.yml

```diff
@@ -17,47 +17,69 @@
     customize_timezone_name: "Hawaiian Standard Time"
     gosc_dns_servers: ['192.168.0.1', '192.168.0.2']
     customize_runonce_echo_string: 'Windows gosc automation test'
-- name: Set fact of the run once command
+
+- name: "Set fact of the run once command"
   ansible.builtin.set_fact:
     customize_runonce: "cmd.exe /c echo {{ customize_runonce_echo_string }} > C:\\gosc_runonce.txt"
 
-- name: Set fact of the VM guest IP before GOSC
+- name: "Set fact of the VM guest IP before GOSC"
   ansible.builtin.set_fact:
     guest_ip_before_gosc: "{{ vm_guest_ip }}"
 
-# Get guest OS hostname from VM guestinfo
-- include_tasks: ../../common/vm_get_config.yml
+- name: "Get guest OS hostname from VM guestinfo"
+  include_tasks: ../../common/vm_get_config.yml
   vars:
     property_list: ['guest.hostName']
 
-- name: Set fact of the hostname before GOSC
+- name: "Set fact of the hostname before GOSC"
   ansible.builtin.set_fact:
     hostname_before_gosc: "{{ vm_config.guest.hostName }}"
+
 - ansible.builtin.debug:
     msg: "Get guest OS hostname/IP before customization: {{ hostname_before_gosc }}/{{ guest_ip_before_gosc }}"
 
-- name: Set fact of default Windows dir
+- name: "Set fact of default Windows dir"
   ansible.builtin.set_fact:
     win_dir: '$env:windir'
-- include_tasks: ../utils/win_get_path.yml
+
+- name: "Get OS directory"
+  include_tasks: ../utils/win_get_path.yml
   vars:
     win_get_path_specified: "{{ win_dir }}"
-- name: Set fact of the absolute path of Windows dir
+
+- name: "Set fact of the absolute path of Windows dir"
   ansible.builtin.set_fact:
     win_dir: "{{ win_get_path_absolute }}"
 - ansible.builtin.debug:
     msg: "Windows GOSC log files in Windows dir: {{ win_dir }}"
 
 # Uninstall OneDrive in Windows 11 for the known 3rd-party issue
 # Paramter 'uninstall_onedrive' is used for internal testing only
-- include_tasks: uninstall_onedrive.yml
+- name: "Uninstall OneDrive"
+  include_tasks: uninstall_onedrive.yml
   when:
     - uninstall_onedrive is defined
     - uninstall_onedrive | bool
     - guest_os_ansible_distribution_ver is version('10.0.22000.0', '>=')
     - guest_os_product_type | lower == 'client'
 
+- name: "Check if there are encrypted volumes in guest OS"
+  include_tasks: ../utils/win_get_bitlocker_volume.yml
+
+- name: "Get BitLocker service status"
+  include_tasks: ../utils/win_get_service_status.yml
+  vars:
+    win_service_name: "BDESVC"
+
+# Disable BitLocker which will cause sysprep failure.
+- name: "Disable Bitlocker"
+  include_tasks: ../utils/win_disable_bitlocker.yml
+  when: >-
+    (bitlocker_volume_list | length > 0) or
+    service_status == "Running"
```

Comment:
service_status and bitlocker_volume_list 
 will be changed in win_disable_bitlocker.yml. It is better not using them at here. I suggest moving checking bitlocker service and volume list into the beginning of win_disable_bitlocker.yml. Firstly, get bitlocker service status, if it is running then disable it.  Then get bitlocker volume list, if not empty, decrypt it.

Reply:
Oh, yes, thanks for the reminder. We should avoid the scenario that the variables in WHEN condition may be changed in the block. I'll re-add the variable need_disable_bitlocker to save the results. 

As to moving them to win_disable_bitlocker.yml, it's not necessary. It's a utility task whose job is just to do disable operation, just as other enable/disable utility jobs. The caller needs to determine if they want to use the utility task.